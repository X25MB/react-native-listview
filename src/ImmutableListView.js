import Immutable from 'immutable';
import React, { PureComponent, PropTypes } from 'react';
import { ListView, InteractionManager } from 'react-native';

import utils from './utils';

/**
 * A ListView capable of displaying {@link https://facebook.github.io/immutable-js/ Immutable} data
 * out of the box.
 */
class ImmutableListView extends PureComponent {

  static propTypes = {
    // Pass through any props that ListView would normally take.
    ...ListView.propTypes,

    // ImmutableListView handles creating the dataSource, so don't pass it in.
    dataSource: PropTypes.oneOf([undefined]),

    // TODO: Allow any type of arbitrary data.
    // eslint-disable-next-line consistent-return
    immutableData: (props, propName, componentName) => {
      // Note: It's not enough to simply validate PropTypes.instanceOf(Immutable.Iterable),
      // because different imports of Immutable.js across files have different class prototypes.
      if (!utils.isImmutableIterable(props[propName])) {
        return new Error(`Invalid prop ${propName} supplied to ${componentName}: Must be Immutable.Iterable.`);
      }
    },

    sectionHeaderHasChanged: PropTypes.func,

    /**
     * How many rows of data to display while waiting for interactions to finish (e.g. Navigation animations).
     * You can use this to improve the animation performance of longer lists when pushing new routes.
     *
     * @see https://facebook.github.io/react-native/docs/performance.html#slow-navigator-transitions
     */
    rowsDuringInteraction: PropTypes.number,
  };

  static defaultProps = {
    ...ListView.getDefaultProps(),

    // The data contained in the section generally doesn't affect the header text, so return false.
    // eslint-disable-next-line no-unused-vars
    sectionHeaderHasChanged: (prevSectionData, nextSectionData) => false,
  };

  state = {
    dataSource: new ListView.DataSource({
      rowHasChanged: (prevRowData, nextRowData) => !Immutable.is(prevRowData, nextRowData),

      getRowData: (dataBlob, sectionID, rowID) => {
        const rowData = utils.getValueFromKey(sectionID, dataBlob);
        return utils.getValueFromKey(rowID, rowData);
      },

      sectionHeaderHasChanged: this.props.sectionHeaderHasChanged,

      getSectionHeaderData: (dataBlob, sectionID) => utils.getValueFromKey(sectionID, dataBlob),
    }),

    interactionOngoing: true,
  };

  componentWillMount() {
    this.canSetState = true;
    this.setStateFromPropsAfterInteraction(this.props);
  }

  componentWillReceiveProps(newProps) {
    this.setStateFromPropsAfterInteraction(newProps);
  }

  componentWillUnmount() {
    this.canSetState = false;
  }

  setStateFromPropsAfterInteraction(props) {
    // Always set state right away before the interaction.
    this.setStateFromProps(props, false);

    // If set, wait for animations etc. to complete before rendering the full list of data.
    if (props.rowsDuringInteraction >= 0) {
      InteractionManager.runAfterInteractions(() => {
        this.setStateFromProps(props, true);
      });
    }
  }

  setStateFromProps(props, interactionHasJustFinished) {
    // In some cases the component will have been unmounted as we receive new props, causing a warning.
    if (!this.canSetState) return;

    const { dataSource, interactionOngoing } = this.state;
    const { immutableData, rowsDuringInteraction, renderSectionHeader } = props;

    const shouldDisplayPartialData = rowsDuringInteraction >= 0 && interactionOngoing && !interactionHasJustFinished;

    const displayData = (shouldDisplayPartialData
      ? immutableData.slice(0, rowsDuringInteraction)
      : immutableData);

    const updatedDataSource = (renderSectionHeader
      ? dataSource.cloneWithRowsAndSections(
          displayData, utils.getKeys(displayData), utils.getRowIdentities(displayData))
      : dataSource.cloneWithRows(
          displayData, utils.getKeys(displayData)));

    this.setState({
      dataSource: updatedDataSource,
      interactionOngoing: interactionHasJustFinished ? false : interactionOngoing,
    });
  }

  render() {
    const { dataSource } = this.state;

    // Note: enableEmptySections is being used to mimic the default behavior of the upcoming version.
    return (
      <ListView
        dataSource={dataSource}
        enableEmptySections
        {...this.props}
      />
    );
  }

}

export default ImmutableListView;
