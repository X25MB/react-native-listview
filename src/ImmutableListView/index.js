import ImmutableListView from './ImmutableListView';

export { ImmutableListView };
export * from './EmptyListView';
