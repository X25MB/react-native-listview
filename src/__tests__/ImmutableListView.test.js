/* eslint-disable import/no-extraneous-dependencies */

import _keys from 'lodash/keys';
import Immutable from 'immutable';
import React from 'react';
import { Text, InteractionManager } from 'react-native';
import renderer from 'react-test-renderer';

import ImmutableListView from '../ImmutableListView';

const EMPTY_DATA = Immutable.List();
const LIST_DATA = Immutable.List(['l', 'i', 's', 't']);
const MAP_DATA = Immutable.Map({ 1: 'm', 2: 'a', 3: 'p' });
const MAP_DATA_COMPLEX = Immutable.Map({ first: ['m', 'a', 'p'], second: ['foo'], third: [], fourth: ['bar'] });

/**
 * @param {*} rowData
 */
function renderRow(rowData) {
  return <Text>{JSON.stringify(rowData)}</Text>;
}

/**
 * @param {Object} sectionData
 * @param {String} category
 */
function renderSectionHeader(sectionData, category) {
  return <Text header>{`${category} (${_keys(sectionData).length} items)`}</Text>;
}

describe('ImmutableListView', () => {
  it('renders with empty data', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={EMPTY_DATA}
        renderRow={renderRow}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders basic List', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={LIST_DATA}
        renderRow={renderRow}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders basic Map', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={MAP_DATA}
        renderRow={renderRow}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders basic Map with section headers', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={MAP_DATA}
        renderRow={renderRow}
        renderSectionHeader={renderSectionHeader}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders more complex Map with section headers', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={MAP_DATA_COMPLEX}
        renderRow={renderRow}
        renderSectionHeader={renderSectionHeader}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ImmutableListView with delayed rendering', () => {
  it('renders basic List during interactions', () => {
    // Mock this method to make sure it's not run.
    InteractionManager.runAfterInteractions = () => {};

    const tree = renderer.create(
      <ImmutableListView
        immutableData={LIST_DATA}
        rowsDuringInteraction={1}
        renderRow={renderRow}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders basic List after interactions', () => {
    // Mock this method to make sure it runs immediately.
    InteractionManager.runAfterInteractions = (callback) => callback();

    const tree = renderer.create(
      <ImmutableListView
        immutableData={LIST_DATA}
        rowsDuringInteraction={1}
        renderRow={renderRow}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
