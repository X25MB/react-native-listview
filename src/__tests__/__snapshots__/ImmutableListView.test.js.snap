exports[`ImmutableListView renders Map: List rows, with section headers 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      first (3 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"m\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"a\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"p\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      second (1 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"foo\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      third (0 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      fourth (1 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"bar\"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: List rows, without section headers 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      [\"m\",\"a\",\"p\"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      [\"foo\"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      []
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      [\"bar\"]
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: Map rows, with section headers 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      first (2 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"data 1\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"data 2\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}>
      second (0 items)
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: Map rows, without section headers 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      {\"row1\":\"data 1\",\"row2\":\"data 2\"}
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      {}
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders basic List 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"lists\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"are\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"great\"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders basic Set 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"one\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"two\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"three\"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders nested List 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      [\"so\",\"are\"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      [\"nested\",\"lists\"]
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders with empty data 1`] = `
<RCTScrollView>
  <View />
</RCTScrollView>
`;

exports[`ImmutableListView with delayed rendering renders basic List after interactions 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"lists\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"are\"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"great\"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with delayed rendering renders basic List during interactions 1`] = `
<RCTScrollView>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      \"lists\"
    </Text>
  </View>
</RCTScrollView>
`;
